packages/connectors-web/src/alchemySmartWallet.ts

```diff
@@ -1,85 +1,333 @@
+     *
+     * TODO: Should we return the same provider instance every time, with that
+     * instance being a proxy object routing to either this or the real provider
+     * if it exists? We would need to do that if Wagmi held on to a provider
+     * that it first got during initialization. It doesn't seem to do that
+     * today, but what if it changed behavior?
+     */
+    const loggedOutProvider: Provider = {
+      on: () => {},
+      removeListener: () => {},
+      request: async (requestParams) => {
+        if (requestParams.method === "eth_accounts") {
+          return [] as any;
+        }
+        const baseProvider = (await innerConnector.getProvider()) as any;
+        return baseProvider.request(requestParams);
+      },
+    };
+
+    let smartWalletClientPromise: Promise<SmartWalletClient> | undefined;
+    let currentChainId: number | undefined;
+    let currentInnerAccount: Address | undefined;
+
+    const resetState = (): void => {
+      smartWalletClientPromise = undefined;
+      currentChainId = undefined;
+      currentInnerAccount = undefined;
+    };
+
+    const createClient = async (): Promise<SmartWalletClient> => {
+      const chainId = await innerConnector.getChainId();
+      const innerAccounts = await innerConnector.getAccounts();
+      if (innerAccounts.length === 0) {
+        throw new Error("No accounts found in base connector");
+      }
+      // TODO: do we always take the first account?
+      const signerAddress = innerAccounts[0];
+      const chain = getChainFromConfig(config, chainId);
+      const provider = (await innerConnector.getProvider()) as EIP1193Provider;
+      const signer = createWalletClient({
+        account: signerAddress,
+        chain,
+        transport: custom(provider),
+      });
+      const smartWalletClient = createSmartWalletClient({
+        signer,
+        transport: alchemyTransport({
+          apiKey: options.apiKey,
+          jwt: options.jwt,
+          url: options.url ?? "https://api.g.alchemy.com/v2",
+        }),
+        chain,
+        policyId: options.policyId,
+        policyIds: options.policyIds,
+      });
+      // TODO: Calling `client.getProvider()` has the side effect of
+      // initializing `client.account` some time after the provider is created.
+      // We should change that so it's explicit and awaitable. Until then,
+      // replicate the logic by hand.
+      const account = await smartWalletClient.requestAccount();
+      smartWalletClient.account = {
+        type: "json-rpc",
+        address: account.address,
+      };
+      return smartWalletClient;
+    };
+
+    const initializeSmartWalletClient = async (): Promise<void> => {
+      smartWalletClientPromise = createClient();
+      const smartWalletClient = await smartWalletClientPromise;
+      currentChainId = smartWalletClient.chain.id;
+      currentInnerAccount = (await innerConnector.getAccounts())[0];
```

Comment:
won't this throw an indexing error if `getAccounts` returns an empty array or is that case impossible?

Reply:
Oh good call. Normally this should always have an account in it since we won't try to initialize without a connected inner connector, but it's very possible that due to race conditions it won't have an account by the time we get here. Let me change that.